## 4.6.2 Task Exercises

### This script includes the following points:

#### 01. Importing Libraries and Loading Files
#### 02. Checking Dataframes
#### 03. Combining Dataframes 
#### 04. Confirming Merge Results
#### 05. Exporting Merged Dataframe

### 01. Importing Libraries and Loading Files

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

# Creating path variable to load files
path = r'C:\Users\widne\Documents\CareerFoundry Exercises\Data_Immersion\Achievement 4\03-2024 Instacart Basket Analysis'

# Reading in orders_products_combined pickle
df_orders_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

# Reading in products_checked_.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

### 02. Checking Dataframes

In [2]:
df_orders_prods_combined.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [9]:
# Dropping _merge
df_orders_prods_combined = df_orders_prods_combined.drop(['_merge'], axis = 1)

In [10]:
df_orders_prods_combined.shape

(32434489, 10)

In [11]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [5]:
# Dropping unnamed columns
df_prods = df_prods.drop(['Unnamed: 0'], axis = 1)
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [6]:
df_prods.shape

(49672, 5)

### 03. Combining Dataframes

The best way to combine the df_orders_prods_combined and the df_prods dataframes is to use the merge function. The two main reasons to combine them this way is that the two dataframes have different shapes, and they both share a common key in the product_id column in both dataframes.

In [12]:
# Testing merge without overwriting
pd.merge(df_orders_prods_combined, df_prods, on = 'product_id', indicator = True) 

order_id  user_id  order_number  order_day_of_week  \
0          2539329        1             1                  2   
1          2398795        1             2                  3   
2           473747        1             3                  3   
3          2254736        1             4                  4   
4           431534        1             5                  4   
...            ...      ...           ...                ...   
32404854   1320836   202557            17                  2   
32404855     31526   202557            18                  5   
32404856    758936   203436             1                  2   
32404857   2745165   203436             2                  3   
32404858   3093936   205420             1                  4   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                         8                     NaN         True         196   
1                         7                    15.0        False         196   
2                        12                    21.0        False         196   
3                         7                    29.0        False         196   
4                        15                    28.0        False         196   
...                     ...                     ...          ...         ...   
32404854                 15                     1.0        False       43553   
32404855                 11                     3.0        False       43553   
32404856                  7                     NaN         True       42338   
32404857                  5                    15.0        False       42338   
32404858                 14                     NaN         True       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices _merge  
0               77              7     9.0   both  
1               77              7     9.0   both  
2               77              7     9.0   both  
3               77              7     9.0   both  
4               77              7     9.0   both  
...            ...            ...     ...    ...  
32404854        64              7     3.7   both  
32404855        64              7     3.7   both  
32404856        50             19     6.9   both  
32404857        50             19     6.9   both  
32404858       130             14    10.3   both  

[32404859 rows x 15 columns]

When testing the merge, the _merge column already exists in the df_orders_prods_combined dataframe, so the column needs to be dropped in order to merge the two dataframes.

In [39]:
# Dropping _merge column in order to merge dataframes
df_orders_prods_combined = df_orders_prods_combined.drop(['_merge'], axis = 1)

# Checking dataframe
df_orders_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered  
0          0  
1          0  
2          0  
3          0  
4          0

In [40]:
# Testing merge without overwriting again
pd.merge(df_orders_prods_combined, df_prods, on = 'product_id', indicator = True) 

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                         8                     NaN         True         196   
1                         7                    15.0        False         196   
2                        12                    21.0        False         196   
3                         7                    29.0        False         196   
4                        15                    28.0        False         196   
...                     ...                     ...          ...         ...   
32404854                 15                     1.0        False       43553   
32404855                 11                     3.0        False       43553   
32404856                  7                     NaN         True       42338   
32404857                  5                    15.0        False       42338   
32404858                 14                     NaN         True       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices _merge  
0               77              7     9.0   both  
1               77              7     9.0   both  
2               77              7     9.0   both  
3               77              7     9.0   both  
4               77              7     9.0   both  
...            ...            ...     ...    ...  
32404854        64              7     3.7   both  
32404855        64              7     3.7   both  
32404856        50             19     6.9   both  
32404857        50             19     6.9   both  
32404858       130             14    10.3   both  

[32404859 rows x 15 columns]

In [13]:
# Merging dataframes into new dataframe
df_ords_prods_merged = df_orders_prods_combined.merge(df_prods, on = 'product_id', indicator = True)

In [14]:
# Checking df_ords_prods_merged dataframe
df_ords_prods_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  
0          0         Soda        77              7     9.0   both  
1          1         Soda        77              7     9.0   both  
2          1         Soda        77              7     9.0   both  
3          1         Soda        77              7     9.0   both  
4          1         Soda        77              7     9.0   both

In [15]:
df_ords_prods_merged.shape

(32404859, 15)

### 04. Confirming Merge Results

In [44]:
df_ords_prods_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

The merged dataframe has 32,404,859 rows that has the data from both dataframes as noted by the both flag.

### 05. Exporting Merged Dataframe

Since this is a large dataframe with many rows, exporting it as a pickle would be best

In [16]:
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))